firedrake/slate/slac/compiler.py

```diff
@@ -0,0 +1,343 @@
+                                 ast.Less(jsym, dof_extent),
+                                 ast.Incr(jsym, 1),
+                                 ast.Assign(ast.Symbol(temp, rank=(tensor_index,)),
+                                            ast.Symbol(builder.coefficient_map[acting_coefficient],
+                                                       rank=(isym, jsym))))
+            # Outer-loop running over node_extent
+            loop = ast.For(ast.Decl("unsigned int", isym, init=0),
+                           ast.Less(isym, node_extent),
+                           ast.Incr(isym, 1),
+                           inner_loop)
+
+            aux_statements.append(loop)
+            aux_temps[acting_coefficient] = temp
+        else:
+            raise NotImplementedError("Auxiliary expression type %s not currently implemented." % type(exp))
+
+    return aux_temps, aux_statements
+
+
+def parenthesize(arg, prec=None, parent=None):
+    """Parenthesizes an expression."""
+    if prec is None or prec >= parent:
+        return arg
+    return "(%s)" % arg
+
+
+def metaphrase_slate_to_cpp(expr, temps, prec=None):
+    """Translates a SLATE expression into its equivalent representation in the Eigen C++ syntax.
+
+    :arg expr: a :class:`slate.TensorBase` expression.
+    :arg temps: a `dict` of temporaries which map a given expression to its corresponding
+                representation as a `coffee.Symbol` object.
+    :arg prec: an argument dictating the order of precedence in the linear algebra operations.
+               This ensures that parentheticals are placed appropriately and the order in which
+               linear algebra operations are performed are correct.
+
+    Returns
+        This function returns a `string` which represents the C/C++ code representation
+        of the `slate.TensorBase` expr.
+    """
+    if isinstance(expr, Tensor):
+        return temps[expr].gencode()
+
+    elif isinstance(expr, Transpose):
+        return "(%s).transpose()" % metaphrase_slate_to_cpp(expr.tensor, temps)
+
+    elif isinstance(expr, Inverse):
+        return "(%s).inverse()" % metaphrase_slate_to_cpp(expr.tensor, temps)
+
+    elif isinstance(expr, Negative):
+        result = "-%s" % metaphrase_slate_to_cpp(expr.tensor, temps, expr.prec)
+        return parenthesize(result, expr.prec, prec)
+
+    elif isinstance(expr, (Add, Sub, Mul)):
+        op = {Add: '+',
+              Sub: '-',
+              Mul: '*'}[type(expr)]
+        result = "%s %s %s" % (metaphrase_slate_to_cpp(expr.operands[0], temps, expr.prec),
+                               op,
+                               metaphrase_slate_to_cpp(expr.operands[1], temps, expr.prec))
+
+        return parenthesize(result, expr.prec, prec)
+
+    elif isinstance(expr, Action):
+        tensor = expr.tensor
+        c = expr._acting_coefficient
+        result = "(%s) * %s" % (metaphrase_slate_to_cpp(tensor, temps, expr.prec), temps[c])
```

Comment:
As I think we discussed in meat-space, it's sort of inconsistent about whether you're using `.tensor` or `.tensors` or `.children`.  But I think fix that it in a separate pass.